### DISCUSSION TITLE
[Share] Customize color labels of reader

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#211] [Action Scripts]
### Description

Customize color labels of reader. Edit the action script below to replace the labels.

> Set the `Event` to `Program Startup`.

<img width="147" alt="image" src="https://github.com/windingwind/zotero-actions-tags/assets/33902321/d4740408-793f-4b8b-800a-f3c00187dcb2">

See also #210

### Event

Program Startup

### Operation

Script

### Data

```javascript
/**
 * Customize color labels of reader
 * @author windingwind
 * @usage Set the `Event` to `Program Startup`; Edit the action script below to replace the labels.
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/211
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/211
 */

// Edit labels below
const replaceLabelMap = {
  "#ffd400": "Custom Yellow",
  "#ff6666": "Custom Red",
  "#5fb236": "Custom Green",
  "#2ea8e5": "Custom Blue",
  "#a28ae5": "Custom Purple",
  "#e56eee": "Custom Magenta",
  "#f19837": "Custom Orange",
  "#aaaaaa": "Custom Grey",
};

function hackContextMenuLabel(event) {
  setTimeout(() => {
    event.reader._iframeWindow?.document
      .querySelectorAll(".context-menu .row")
      .forEach((e) => {
        const color = e.querySelector("path[fill]")?.getAttribute("fill");
        if (!color) {
          return;
        }
        if (color in replaceLabelMap) {
          e.innerHTML =
            e.querySelector("svg")?.outerHTML + replaceLabelMap[color];
        }
      });
  }, 10);
}

Zotero.Reader.registerEventListener(
  "createAnnotationContextMenu",
  hackContextMenuLabel,
  "zoterotag@euclpts.com"
);

Zotero.Reader.registerEventListener(
  "createColorContextMenu",
  hackContextMenuLabel,
  "zoterotag@euclpts.com"
);
```

**Comment by OpenFoam-User** [comment]
Very nice! thank you very much for this script.

**Comment by Geo123abc** [comment]
选中一个条目，shift+enter 在一个独立窗口中打开PDF，在这个独立窗口中动作无效，不知道是我姿势不对？还是？

**Reply by windingwind** [reply]
我确认该动作没问题。如果认为是bug请提交issue

**Reply by Geo123abc** [reply]
在独立窗口中也有效吗？

**Reply by windingwind** [reply]
对

**Reply by Geo123abc** [reply]
我又试了一下，确实可以，**但是颜色似乎不能自定义**，我在想，颜色能不能也支持自定义？就类似zotero style的那个更改高亮的功能（zotero style可以在主窗口阅读时更改颜色，但是在独立窗口中的阅读器中无效）

**Comment by mjthoraval** [comment]
I am not able to get this script to work.

Here are the parameters I have set:
![image](https://github.com/windingwind/zotero-actions-tags/assets/36054098/8662bde6-248d-42af-8e0e-b525226fa7b5)

But after restart, I still get:
![image](https://github.com/windingwind/zotero-actions-tags/assets/36054098/1cb29486-593e-4a78-b803-c2ed495f844c)
![image](https://github.com/windingwind/zotero-actions-tags/assets/36054098/e2a4abc5-cb98-4172-ae8e-f1dcc85bd1a4)

Restart with Logging Enabled: D762267552

Just to confirm also: will the labels return to the default labels when I disable this script and restart Zotero?

**Reply by wakewon** [reply]
I suspect this script is no longer available under the current version of Zotero.

**Reply by mjthoraval** [reply]
For reference: https://forums.zotero.org/discussion/103437/feature-request-labelling-the-highlight-text-colours/

**Reply by windingwind** [reply]
updated

**Reply by mjthoraval** [reply]
Thank you very much! I confirm that it is working nicely now.

Just a minor problem: it seems that the space between the color square and the label is missing:
![image](https://github.com/user-attachments/assets/b2a4785c-305e-4a27-ba80-a1a6d5550297)

Even adding a space manually does not work:
![image](https://github.com/user-attachments/assets/8374da95-329a-4f0f-85a6-0b2cc1dc912b)
![image](https://github.com/user-attachments/assets/007231df-2fe9-40c9-934a-a30a5fe6f515)

See the difference with the default labels:
![image](https://github.com/user-attachments/assets/6f5b1cf9-962d-4b8c-ae81-91640e3488b8)

Zotero 7.0.0-beta.117+794dc6bf0 (64-bit)
Windows 10

**Reply by ramonmi** [reply]
This script seems to not work in the popup after selecting text. The tooltip text remains the default color label.

**Reply by weihv** [reply]
Please check here: https://github.com/windingwind/zotero-actions-tags/discussions/211?sort=new#discussioncomment-10564203

**Reply by ramonmi** [reply]
Fantastic! Thank you very much.

**Reply by vermavinay982** [reply]
> Thank you very much! I confirm that it is working nicely now.
> 
> Just a minor problem: it seems that the space between the color square and the label is missing: ![image](https://private-user-images.githubusercontent.com/36054098/356883165-b2a4785c-305e-4a27-ba80-a1a6d5550297.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAwMzMyMzUsIm5iZiI6MTc2MDAzMjkzNSwicGF0aCI6Ii8zNjA1NDA5OC8zNTY4ODMxNjUtYjJhNDc4NWMtMzA1ZS00YTI3LWJhODAtYTFhNmQ1NTUwMjk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA5VDE4MDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxZmZiYWU3ZTA4YzRhMDRmY2RjZjY0Mzk3Mzc4NDhhMGMxZjlmMTgzYzM5MDgxNGQyODlhODcwM2NmOWU1ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.haA47bu-KzNE-AKww3x6wvB08qRWEkBOm6w9Pb_CXbI)
> 
> Even adding a space manually does not work: ![image](https://private-user-images.githubusercontent.com/36054098/356883199-8374da95-329a-4f0f-85a6-0b2cc1dc912b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAwMzMyMzUsIm5iZiI6MTc2MDAzMjkzNSwicGF0aCI6Ii8zNjA1NDA5OC8zNTY4ODMxOTktODM3NGRhOTUtMzI5YS00ZjBmLTg1YTYtMGIyY2MxZGM5MTJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA5VDE4MDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0MDZhYzZkZWU5NmNmM2Q3YjJhZjJjNzFkZTc3ODlhYzliNWRjYjRiMDdlN2JkMmIzZDc3ZmUwMjgyODVhYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._SNOrfTjYUZEgEYn2ZFhInD1xZ3gzKrG8GxWwsjJeOo) ![image](https://private-user-images.githubusercontent.com/36054098/356883176-007231df-2fe9-40c9-934a-a30a5fe6f515.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAwMzMyMzUsIm5iZiI6MTc2MDAzMjkzNSwicGF0aCI6Ii8zNjA1NDA5OC8zNTY4ODMxNzYtMDA3MjMxZGYtMmZlOS00MGM5LTkzNGEtYTMwYTVmZTZmNTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA5VDE4MDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNTUxOGUyYzUzMDk2ZTI2Njc3NTg5OTIxMjkwMTFiZGU3YTdmZGMyNmU5ZWQzOTdkNzM5YTBlMWY0ODNjMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.q9S_ofBiGFfUKY_UgiKd4FUxQ1-zjgis4hTXdVmGxFg)
> 
> See the difference with the default labels: ![image](https://private-user-images.githubusercontent.com/36054098/356883236-6f5b1cf9-962d-4b8c-ae81-91640e3488b8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAwMzMyMzUsIm5iZiI6MTc2MDAzMjkzNSwicGF0aCI6Ii8zNjA1NDA5OC8zNTY4ODMyMzYtNmY1YjFjZjktOTYyZC00YjhjLWFlODEtOTE2NDBlMzQ4OGI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA5VDE4MDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMDk0M2ExZmVjNDBlZWJlY2Y3NGVlOTZiOGZlZGRkODA1OTg0MzMwOTcyY2M5NzA1OTM2Y2ZlZGJjOGVhOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DEk9YT00HkqqczBYd3gZiaO4W2WzVZm_98PdTRC0Gtw)
> 
> Zotero 7.0.0-beta.117+794dc6bf0 (64-bit) Windows 10

Now you can have spaces as well, it looks cleaner now. 

```js
const replaceLabelMap = {
  "#ffd400": " Facts",
  "#ff6666": " Custom Red",
  "#5fb236": " Results",
  "#2ea8e5": " Definitions",
  "#a28ae5": " Methodology",
  "#e56eee": " Future Plans",
  "#f19837": " References",
  "#aaaaaa": " Custom Grey",
};

function hackContextMenuLabel(event) {
  setTimeout(() => {
    event.reader._iframeWindow?.document
      .querySelectorAll(".context-menu .row")
      .forEach((e) => {
        const color = e.querySelector("path[fill]")?.getAttribute("fill");
        if (!color) return;
        if (color in replaceLabelMap) {
          const icon = e.querySelector("svg")?.outerHTML;
          const label = replaceLabelMap[color];
          // Add a proper space or thin space between icon and label
          e.innerHTML = `${icon}&nbsp;${label.trim()}`;
        }
      });
  }, 10);
}
```

**Comment by jmw999** [comment]
It doesn't seem to work anymore for me either - but such a useful addition! Any ideas?

**Reply by mjthoraval** [reply]
Have you updated the script to the latest version (modified in the post at the top)?
Can you give more details on your OS, Zotero version, plugin version, and the steps you have followed?

**Reply by jmw999** [reply]
I'm on version 2.0.0 of the plug in,  7.0.1-beta.3+ec7ffa8c0 (64-bit) of Zotero, Windows 10 pro, version 22H2. I installed the plug in yesterday, copied the script from the plug in this morning, and pasted it into the Monaco box loaded by the plugin. the rest of the settings are as per the screenshot.  The only other changes I made were to edit lines 56-65 as follows:
 // Edit labels below
const replaceLabelMap = {
  "#ffd400": "Yellow - Useful",
  "#ff6666": "Red - Definition",
  "#5fb236": "Green - Quote",
  "#2ea8e5": "Blue - Good point",
  "#a28ae5": "Custom Purple",
  "#e56eee": "Magenta - Counterargument",
  "#f19837": "Custom Orange",
  "#aaaaaa": "Custom Grey",
};
![image](https://github.com/user-attachments/assets/76ae7083-5c5a-4f72-aea8-996acab43ae4)

Hoping I have missed out a blindingly obvious step - as this adaptation seems so needed.

**Reply by mjthoraval** [reply]
Does the original script work before making any modification?

> The only other changes I made were to edit lines 56-65 as follows:

You should just replace lines 10-19.
That is working for me:
![image](https://github.com/user-attachments/assets/c0533507-3290-4a0a-bbbc-9d31a853a897)

**Comment by jmw999** [comment]
Just to add that I've also edited lines 10-19 as well - but it didn't make any difference when I restarted Zotero.

**Comment by jmw999** [comment]
I really appreciate you taking the time to respond but this is what I get after editing 10-19 but returning 56-65 to what they were. 
![image](https://github.com/user-attachments/assets/6e7b0f3f-dec2-415e-bea3-a61a0d06ccab)

**Reply by mjthoraval** [reply]
I am a bit confused here.
The code shared here only has 48 lines for me:
![image](https://github.com/user-attachments/assets/eceedae9-a6eb-4f31-8943-ebf6a03ac0e9)

Can you show a screenshot of the code you have?

**Comment by jmw999** [comment]
Interesting! Mine is much longer - 95 lines :(  I'll try to print it out and get rid of the surplus.  That's very bizarre, as I just copied it from the data bit further up the thread.. Thank you so much though :)

**Comment by jmw999** [comment]
Cracked it - thank you! I've no idea how that went wrong, but it did.

**Comment by weihv** [comment]
Add customize color labels to selection popup and sidebar tag, add space to original context menus

```javascript
/**
 * Customize color labels of reader
 * @author windingwind
 * @editor weihv
 * @eidt Add customize color labels to selection popup and sidebar tag, add space to original context menus
 * @reference https://www.zotero.org/support/dev/zotero_7_for_developers#custom_reader_event_handlers
 * @usage Set the Event to Program Startup; Edit the action script below to replace the labels.
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/211
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/211
 */

// Edit labels below
const replaceLabelMap = {
  "#ffd400": "Custom Yellow",
  "#ff6666": "Custom Red",
  "#5fb236": "Custom Green",
  "#2ea8e5": "Custom Blue",
  "#a28ae5": "Custom Purple",
  "#e56eee": "Custom Magenta",
  "#f19837": "Custom Orange",
  "#aaaaaa": "Custom Grey",
};

function hackContextMenuLabel(event) {
  setTimeout(() => {
    // Check if we are dealing with a context menu
    if (event.reader._iframeWindow?.document.querySelector(".context-menu")) {
      // This is the context menu case
      event.reader._iframeWindow?.document
        .querySelectorAll(".context-menu .row")
        .forEach((e) => {
          const color = e.querySelector("path[fill]")?.getAttribute("fill");
          if (!color) {
            return;
          }
          if (color in replaceLabelMap) {
            e.innerHTML = e.querySelector("svg")?.outerHTML + "&nbsp;" + replaceLabelMap[color];
          }
        });
    }
  }, 10);
}

function hackTextSelectionPopup(event) {
  setTimeout(() => {
    // Target the text selection popup
    const popup = event.reader._iframeWindow?.document.querySelector(".selection-popup");
    
    if (popup) {
      // Modify the title of each button
      popup.querySelectorAll(".toolbar-button.color-button").forEach((button) => {
        const svgElement = button.querySelector("path[fill]");
        if (svgElement) {
          const color = svgElement.getAttribute("fill");
          if (color && replaceLabelMap[color]) {
            button.setAttribute("title", replaceLabelMap[color]);
          }
        }
      });
    }
  }, 10);
}

function hackSidebarAnnotationHeader(event) {
  setTimeout(() => {
	// Target sidebar bottom selector
    const selectorMenu = event.reader._iframeWindow?.document.querySelector(".selector");
    if (selectorMenu) {
      selectorMenu.querySelectorAll(".color").forEach((button) => {
        const svgElement = button.querySelector("path[fill]");
        if (svgElement) {
          const color = svgElement.getAttribute("fill");
          if (color && replaceLabelMap[color]) {
            // first line shows label name on color tag
            // second line show label name after color tag
            button.setAttribute("title", replaceLabelMap[color]); // comment this line if want to show color name on color tag 
            button.innerHTML = svgElement.closest("svg")?.outerHTML + "&nbsp;" + replaceLabelMap[color];
          }
        }
      });
    }
  }, 10);
}


// Register event listeners for context menus, popup, selector
// For left sidebar's annotation right-click menu
Zotero.Reader.registerEventListener(
  "createAnnotationContextMenu",
  hackContextMenuLabel,
  "zoterotag@euclpts.com"
);

// For top toolbar's color picker menu
Zotero.Reader.registerEventListener(
  "createColorContextMenu",
  hackContextMenuLabel,
  "zoterotag@euclpts.com"
);

// For reader text selection popup
Zotero.Reader.registerEventListener(
  "renderTextSelectionPopup",
  hackTextSelectionPopup,
  "zoterotag@euclpts.com"
);

// For left sidebar's annotation header (bottom left tag filtering)
Zotero.Reader.registerEventListener(
  "renderSidebarAnnotationHeader",
  hackSidebarAnnotationHeader,
  "zoterotag@euclpts.com"
);
```

**Reply by physicslog** [reply]
Btw can you also add the sorting functionality based on the position of keys in `replaceLabelMap` ?

**Reply by mjthoraval** [reply]
Thank you very much for this update. It is working great.
I have just changed the space from `"&nbsp;"` to `"&ensp;"` to make it a bit longer.
`"&nbsp;"`:
![image](https://github.com/user-attachments/assets/e113e6f8-1dde-4deb-98f2-300727dce164)
`"&ensp;"`:
![image](https://github.com/user-attachments/assets/3675586e-2717-4ad0-8701-a8bc399ae78a)

But it seems that the space is still not the same as the original padding?
Original:
![image](https://github.com/user-attachments/assets/8ba105f7-66e6-4757-94c8-4137dba17c47)

**Reply by weihv** [reply]
> Btw can you also add the sorting functionality based on the position of keys in `replaceLabelMap` ?

Can u explain what does this means?

**Reply by weihv** [reply]
> Thank you very much for this update. It is working great. I have just changed the space from `"&nbsp;"` to `"&ensp;"` to make it a bit longer. `"&nbsp;"`: ![image](https://private-user-images.githubusercontent.com/36054098/365545802-e113e6f8-1dde-4deb-98f2-300727dce164.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5Nzg2NTUsIm5iZiI6MTcyNTk3ODM1NSwicGF0aCI6Ii8zNjA1NDA5OC8zNjU1NDU4MDItZTExM2U2ZjgtMWRkZS00ZGViLTk4ZjItMzAwNzI3ZGNlMTY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDE0MjU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlOTQ0YTBhZTA2YTA4MDY5MGI2ODIwYWViMjJjMDgyYmFmMGEzYmU1NTFiMjg1YzliZjY0ZWQ1ZTZkOTYwMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9rFK89gOtIS6VwHznlFd5IcaqINi-N32XLdQ-lU9qHQ) `"&ensp;"`: ![image](https://private-user-images.githubusercontent.com/36054098/365546157-3675586e-2717-4ad0-8701-a8bc399ae78a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5Nzg2NTUsIm5iZiI6MTcyNTk3ODM1NSwicGF0aCI6Ii8zNjA1NDA5OC8zNjU1NDYxNTctMzY3NTU4NmUtMjcxNy00YWQwLTg3MDEtYThiYzM5OWFlNzhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDE0MjU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwMDZmMGJhNjNlN2RiZjJhNmEwOWIzZDBjODc4MzE4NmVjMzU5YmVhN2JjZjdjM2Y0MTYzNjFkYmI1YTlmZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xkBUjRRtqz8lCYAegeYv1hvlQrXEoyFhlFw0xODt69o)
> 
> But it seems that the space is still not the same as the original padding? Original: ![image](https://private-user-images.githubusercontent.com/36054098/365546532-8ba105f7-66e6-4757-94c8-4137dba17c47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU5Nzg2NTUsIm5iZiI6MTcyNTk3ODM1NSwicGF0aCI6Ii8zNjA1NDA5OC8zNjU1NDY1MzItOGJhMTA1ZjctNjZlNi00NzU3LTk0YzgtNDEzN2RiYTE3YzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEwVDE0MjU1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA1NDNiYjU3MmZlNWM2YzhkZmFmZjFlNDU2Zjc2OGVkMTI1NGFmNzE3Y2E0M2ViZmJiMDIxZGIwM2U1OWZlNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._UCJHkvO8sVX3nvz_qy0GhlBcbQQRsIfPGTqEvPcHhM)

I pasted the original button item here:
```html
<button tabindex="-1" class="row basic"><div class="icon"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" fill="none"><path d="M1 3C1 1.89543 1.89543 1 3 1H13C14.1046 1 15 1.89543 15 3V13C15 14.1046 14.1046 15 13 15H3C1.89543 15 1 14.1046 1 13V3Z" fill="#ff6666"></path><path d="M1.5 3C1.5 2.17157 2.17157 1.5 3 1.5H13C13.8284 1.5 14.5 2.17157 14.5 3V13C14.5 13.8284 13.8284 14.5 13 14.5H3C2.17157 14.5 1.5 13.8284 1.5 13V3Z" stroke="black" stroke-opacity="0.1"></path></svg></div>Red</button>
```

Depends on where you put text. Inside or out side <div class="icon"></div>

**Reply by physicslog** [reply]
Sure! For example:
```js
const replaceLabelMap = {
  "#ff6666": "Custom Red",  // order is changed
  "#ffd400": "Custom Yellow",  // order is changed
  "#2ea8e5": "Custom Blue",  // order is changed 
  "#5fb236": "Custom Green", // order is changed
  "#a28ae5": "Custom Purple",
  "#e56eee": "Custom Magenta",
  "#f19837": "Custom Orange",
  "#aaaaaa": "Custom Grey",
};
```
Notice that the position of colors in `replaceLabelMap` are in different order than the Zotero's label map. So, after the relabeled, it would be great if you ordered the label based on the `replaceLabelMap`.

**Reply by weihv** [reply]
> Sure! For example:
> 
> ```js
> const replaceLabelMap = {
>   "#ff6666": "Custom Red",  // order is changed
>   "#ffd400": "Custom Yellow",  // order is changed
>   "#2ea8e5": "Custom Blue",  // order is changed 
>   "#5fb236": "Custom Green", // order is changed
>   "#a28ae5": "Custom Purple",
>   "#e56eee": "Custom Magenta",
>   "#f19837": "Custom Orange",
>   "#aaaaaa": "Custom Grey",
> };
> ```
> 
> Notice that the position of colors in `replaceLabelMap` are in different order than the Zotero's label map. So, after the relabeled, it would be great if you ordered the label based on the `replaceLabelMap`.

JavaScript objects like replaceLabelMap here do not have a guaranteed order even if you can see order here.

**Reply by physicslog** [reply]
Yep that's right. But able to have the order would be great!

**Reply by physicslog** [reply]
I modified only the `hackContextMenuLabel`. It reorders the colour label based on the position of `replaceLabelMap` and it also reorders for popups but in a weird way. Maybe there should be a good way to do the same thing :) 

```js
function hackContextMenuLabel(event) {
  setTimeout(() => {
    const contextMenu = event.reader._iframeWindow?.document.querySelector(".context-menu");
    if (contextMenu) {
      const rows = Array.from(contextMenu.querySelectorAll(".row"));
      const colorOrder = Object.keys(replaceLabelMap);

      const orderedRows = rows.sort((a, b) => {
        const colorA = a.querySelector("path[fill]")?.getAttribute("fill");
        const colorB = b.querySelector("path[fill]")?.getAttribute("fill");
        return colorOrder.indexOf(colorA) - colorOrder.indexOf(colorB);
      });

      orderedRows.forEach((e) => {
        const color = e.querySelector("path[fill]")?.getAttribute("fill");
        if (color && replaceLabelMap[color]) {
          e.innerHTML = e.querySelector("svg")?.outerHTML + "&nbsp;" + replaceLabelMap[color];
        }
        contextMenu.appendChild(e);
      });
    }
  }, 10);
}
```

**Comment by OESNES** [comment]
Just learned about scripts and found this one I have wanted for a while.  This is an awesome script. Thanks
